src/hardware/input/intel8042.cpp

```diff
@@ -0,0 +1,1044 @@
+
+static size_t buffer_start_idx = 0;
+static size_t buffer_num_used  = 0;
+
+static size_t waiting_bytes_from_aux = 0;
+static size_t waiting_bytes_from_kbd = 0;
+
+// true = delay timer is in progress
+static bool delay_running = false;
+// true = delay timer expired, event can be sent immediately
+static bool delay_finished = true;
+
+// Executing command, do not notify devices about readiness for accepting frame
+static bool should_skip_device_notify = false;
+
+// Command currently being executed, waiting for parameter
+static Command current_command = Command::None;
+
+// ***************************************************************************
+// Helper routines to log various warnings
+// ***************************************************************************
+
+static void warn_buffer_full()
+{
+	static constexpr uint32_t threshold_ms = 15 * 1000; // 15 seconds
+
+	static bool already_warned = false;
+	static uint32_t last_timestamp = 0;
+
+	if (!already_warned || (PIC_Ticks - last_timestamp > threshold_ms)) {
+		LOG_WARNING("I8042: Internal buffer overflow");
+		last_timestamp = PIC_Ticks;
+		already_warned = true;
+	}
+}
+
+static void warn_controller_mode()
+{
+	static bool already_warned = false;
+	if (!already_warned) {
+		LOG_WARNING("I8042: Switching controller to AT mode not emulated");
+		already_warned = true;
+	}
+}
+
+static void warn_internal_ram_access()
+{
+	static bool already_warned = false;
+	if (!already_warned) {
+		LOG_WARNING("I8042: Accessing internal RAM (other than byte 0x00) gives vendor-specific results");
+		already_warned = true;
+	}
+}
+
+static void warn_line_pulse()
+{
+	static bool already_warned = false;
+	if (!already_warned) {
+		LOG_WARNING("I8042: Pulsing line other than RESET not emulated");
+		already_warned = true;
+	}
+}
+
+static void warn_read_test_inputs()
+{
+	static bool already_warned = false;
+	if (!already_warned) {
+		LOG_WARNING("I8042: Reading test inputs not implemented");
+		already_warned = true;
+	}
+}
+
+static void warn_vendor_lines()
+{
+	static bool already_warned = false;
+	if (!already_warned) {
+		LOG_WARNING("I8042: No vendor-specific commands to manipulate controller lines are emulated");
+		already_warned = true;
+	}
+}
+
+static void warn_unknown_command(const Command command)
+{
+	static bool already_warned[UINT8_MAX + 1];
+	const uint8_t code = static_cast<uint8_t>(command);
+	if (!already_warned[code]) {
+		LOG_WARNING("I8042: Unknown command 0x%02x", code);
+		already_warned[code] = true;
+	}
+}
+
+// ***************************************************************************
+// XT translation for keyboard input
+// ***************************************************************************
+
+static uint8_t get_translated(const uint8_t byte)
+{
+	// A drain bamaged keyboard input translation
+
+	// Values taken from 86Box source code,
+	// can also be found in many other places
+
+	// clang-format off
+	static const uint8_t translation_table[] = {
```

Comment:
`static constexpr` -> might as well initialize this static array at compile (once and for all) instead each time on startup.

Reply:
OK, changed.